<a href="https://colab.research.google.com/github/MiguelSanz2/APRENDIZAJE_AUTOMATICO_23_24/blob/main/Practica3_CNN.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>





# PRÁCTICA 3:

REDES DE NEURONAS CONVOLUCIONALES



---



**Grupo 07:**   *Componentes:*

*   Carlos Adaro Cacho
*   Juan Rivera Sánchez
*   Miguel Sanz Almau



---






**Introducción**

En este código, se muestra cómo utilizamos TensorFlow para implementar una red neuronal convolucional (CNN) para la clasificación de imágenes utilizando el conjunto de datos CIFAR-100. La CNN es una arquitectura de red neuronal especialmente diseñada para procesar datos de imágenes y ha demostrado ser muy efectiva en tareas de clasificación de imágenes.

In [4]:
import tensorflow as tf
from tensorflow.keras.datasets import cifar100
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout, BatchNormalization
from tensorflow.keras.callbacks import LearningRateScheduler

**Paso 1:**

Descargar el conjunto de datos CIFAR-100
En este paso, descargamos el conjunto de datos CIFAR-100 utilizando la función cifar100.load_data(). El conjunto de datos se divide en datos de entrenamiento y datos de prueba, que se almacenan en las variables x_train, y_train, x_test y y_test.

In [5]:
(x_train, y_train), (x_test, y_test) = cifar100.load_data()

**Paso 2:**

Preprocesar los datos
En este paso, realizamos el preprocesamiento de los datos dividiendo los valores de los píxeles entre 255.0 para normalizarlos en el rango de 0 a 1. Esto ayudará a mejorar la convergencia del modelo durante el entrenamiento.

In [6]:
x_train = x_train / 255.0
x_test = x_test / 255.0

**Paso 3:**

Definir la estructura de la red neuronal convolucional
En este paso, se define la estructura de la red neuronal convolucional utilizando la clase Sequential de TensorFlow. Se agregan capas convolucionales, capas de agrupación máxima, capas de aplanamiento y capas completamente conectadas para construir la arquitectura de la CNN.

In [7]:
model = Sequential()
model.add(Conv2D(32, (3, 3), activation='relu', padding='same', input_shape=(32, 32, 3)))
model.add(BatchNormalization())
model.add(Conv2D(32, (3, 3), activation='relu', padding='same'))
model.add(BatchNormalization())
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.5))

model.add(Conv2D(64, (3, 3), activation='relu', padding='same'))
model.add(BatchNormalization())
model.add(Conv2D(64, (3, 3), activation='relu', padding='same'))
model.add(BatchNormalization())
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.5))

model.add(Flatten())
model.add(Dense(512, activation='relu'))
model.add(BatchNormalization())
model.add(Dropout(0.5))
model.add(Dense(100, activation='softmax'))

**Paso 4:**

Compilar el modelo
En este paso, se compila el modelo especificando el optimizador y la función de pérdida. En este caso, se utiliza el optimizador Adam y la función de pérdida de entropía cruzada escasa.

In [8]:
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])

**Paso 5:**

Programación del aprendizaje: El código define una función lr_schedule que ajusta la tasa de aprendizaje durante el entrenamiento. Esta función se utiliza como argumento en el callback LearningRateScheduler, que ajusta la tasa de aprendizaje según el número de épocas.

In [9]:
def lr_schedule(epoch):
    lr = 1e-3
    if epoch > 10:
        lr *= 0.4
    elif epoch > 25:
        lr *= 0.2
    return lr

lr_scheduler = LearningRateScheduler(lr_schedule)

history = model.fit(x_train, y_train, batch_size=128, epochs=30, validation_data=(x_test, y_test), callbacks=[lr_scheduler])

Epoch 1/30
391/391 [==============================] - 11s 21ms/step - loss: 4.1438 - accuracy: 0.1152 - val_loss: 4.3750 - val_accuracy: 0.0771 - lr: 0.0010
Epoch 2/30
391/391 [==============================] - 8s 21ms/step - loss: 3.2366 - accuracy: 0.2284 - val_loss: 2.8210 - val_accuracy: 0.3028 - lr: 0.0010
Epoch 3/30
391/391 [==============================] - 8s 20ms/step - loss: 2.7864 - accuracy: 0.3025 - val_loss: 2.5067 - val_accuracy: 0.3618 - lr: 0.0010
Epoch 4/30
391/391 [==============================] - 8s 20ms/step - loss: 2.5263 - accuracy: 0.3530 - val_loss: 2.2944 - val_accuracy: 0.4068 - lr: 0.0010
Epoch 5/30
391/391 [==============================] - 8s 21ms/step - loss: 2.3569 - accuracy: 0.3902 - val_loss: 2.1418 - val_accuracy: 0.4376 - lr: 0.0010
Epoch 6/30
391/391 [==============================] - 8s 19ms/step - loss: 2.2381 - accuracy: 0.4135 - val_loss: 2.0495 - val_accuracy: 0.4601 - lr: 0.0010
Epoch 7/30
391/391 [==============================] - 8s 20ms/s

**Paso 6:**

Evaluar el modelo
En este paso, se evalúa el rendimiento del modelo utilizando los datos de prueba. Se calcula la pérdida y la precisión del modelo en los datos de prueba.

In [10]:
test_loss, test_acc = model.evaluate(x_test, y_test, verbose=2)

print('Precisión en los datos de prueba:', test_acc)

313/313 - 1s - loss: 1.6561 - accuracy: 0.5555 - 1s/epoch - 3ms/step
Precisión en los datos de prueba: 0.5554999709129333


**Paso 7:**

Realizar predicciones
En este paso, se realiza una predicción utilizando el modelo entrenado en una imagen seleccionada al azar del conjunto de datos de prueba. Se muestra la etiqueta verdadera y la etiqueta predicha para compararlas.

In [11]:
predictions = model.predict(x_test)
model.summary()

import numpy as np

# Seleccionar una imagen al azar del conjunto de datos de prueba
random_index = np.random.randint(0, len(x_test))
random_image = x_test[random_index]
true_label = y_test[random_index]

# Realizar la predicción con el modelo
predicted_label = np.argmax(model.predict(np.expand_dims(random_image, axis=0)), axis=-1)

# Mostrar la imagen y su etiqueta
print(f'Etiqueta verdadera: {true_label}')
print(f'Etiqueta predicha: {predicted_label}')

313/313 [==============================] - 1s 3ms/step
Model: "sequential_3"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d_12 (Conv2D)          (None, 32, 32, 32)        896       
                                                                 
 batch_normalization_15 (Ba  (None, 32, 32, 32)        128       
 tchNormalization)                                               
                                                                 
 conv2d_13 (Conv2D)          (None, 32, 32, 32)        9248      
                                                                 
 batch_normalization_16 (Ba  (None, 32, 32, 32)        128       
 tchNormalization)                                               
                                                                 
 max_pooling2d_6 (MaxPoolin  (None, 16, 16, 32)        0         
 g2D)                                                            

1. **Explicación del model.summary()**

* El entrenamiento se llevó a cabo durante 30 épocas (Epoch 1/30 hasta Epoch 30/30).

* Se realizaron 391 iteraciones (batches) por época.

* El tiempo de ejecución por iteración fue de alrededor de 8-12 segundos por lote (step).

* La función de pérdida (loss) en el conjunto de entrenamiento disminuyó de 4.0970 a 1.3304, lo cual indica que el modelo mejoró en la capacidad de hacer predicciones conforme avanzaba el entrenamiento.

* La precisión (accuracy) en el conjunto de entrenamiento aumentó de 0.1208 a 0.6185.

* Se proporcionan métricas similares para el conjunto de validación (val_loss y val_accuracy).

2. **Optimizador y Tasa de Aprendizaje (LearningRateScheduler):**

* El optimizador utilizado parece ser un optimizador basado en gradiente descendente con una tasa de aprendizaje inicial de 0.001.

* A partir de la Epoch 12/30, la tasa de aprendizaje disminuyó a 0.0004.

3. **Descripción del Modelo:**

* El modelo es de tipo secuencial (Sequential) y tiene varias capas, incluyendo capas convolucionales (Conv2D), capas de normalización (BatchNormalization), capas de agrupación máxima (MaxPooling2D), capas de abandono (Dropout), y capas densas (Dense).

* La última capa densa tiene 100 unidades de salida, lo que sugiere que este modelo se entrenó para clasificar imágenes en 100 clases diferentes.

4. **Parámetros del Modelo:**

* El modelo tiene un total de 2,217,348 parámetros, de los cuales 2,215,940 son entrenables y 1,408 son no entrenables.

En resumen, el modelo fue entrenado para la clasificación de imágenes utilizando el algoritmo de retropropagación (backpropagation) a través de 30 épocas, y parece haber mejorado en términos de pérdida y precisión a lo largo del tiempo. La disminución de la tasa de aprendizaje aplicando (LearningRateScheduler) después de la Epoch 12/30 sugiere que se puede estar utilizando un esquema de ajuste de tasa de aprendizaje para una convergencia más efectiva.





**Conceptos clave**


Conjunto de datos CIFAR-100: El conjunto de datos CIFAR-100 es un conjunto de imágenes de 32x32 píxeles que se divide en 100 clases diferentes. Cada imagen está etiquetada con una de las 100 clases.

Preprocesamiento de datos: Antes de alimentar los datos a la red neuronal, es común realizar un preprocesamiento para normalizar los valores de los píxeles y mejorar la convergencia del modelo.

Estructura de la red neuronal convolucional: La CNN utilizada en este código consta de varias capas convolucionales, capas de agrupación máxima, capas de aplanamiento y capas completamente conectadas. Estas capas se combinan para extraer características de las imágenes y realizar la clasificación.

Compilación y entrenamiento del modelo: Después de definir la estructura de la red neuronal, se compila el modelo especificando el optimizador y la función de pérdida. Luego, el modelo se entrena utilizando los datos de entrenamiento y se valida utilizando los datos de prueba.

Evaluación y predicción del modelo: Una vez entrenado el modelo, se evalúa su rendimiento utilizando los datos de prueba. También se pueden realizar predicciones utilizando el modelo entrenado en nuevas imágenes.

**Gestión del grupo**

Durante el trabajo realizado por el grupo hemos primero programado el código de la red neuronal convolucional y tras ello, hemos ido entrenando por separado la red modificando diferentes elementos del código así como la tasa de aprendizaje, el Dropout, número de capas y épocas,etc.Trabajando para conseguir un aumento de la precisión (‘val_accuracy’) llegamos a obtener entre 0.55-0.56,

A partir de ese resultado hemos conseguido que la etiqueta verdadera coincida con la clase predicha.
Para la memoria nos hemos dividido la implementación de los comentarios en: introducción, pasos del código, conceptos clave y explicación de nuestro ‘model.summary()’.